README.md
# Yorgey CIS 194
u

Chris Allen (a.k.a. [bytemyapp][b]), one of the authors of [Haskell
Book][hbook], wrote an [article][htlh] about his recommended path to _learn
Haskell_.

Between others online resources, he mentions [Brent Yorgey][yorgey]'s course
_CIS 194_, available [online][cis194].

> Do _this first_ if aren't getting the Haskell Book, this is the best _free_
  introduction to Haskell.

In this [repository][repo], I will try follow this course and the proposed exercises.

---
[b]: https://github.com/bitemyapp
[hbook]: http://haskellbook.com/
[htlh]: https://github.com/bitemyapp/learnhaskell
[yorgey]: https://byorgey.wordpress.com/
[cis194]: http://www.seas.upenn.edu/~cis194/spring13/
[repo]: https://github.com/erlnow/yorgey-cis194

## Test

TODO: write about testing with Hspec, QuickCheck and doctest and
the error about ">prop" with doctest and how to solve it, if finally i can

~~~~
> --
> -- $
> -- prop> sumtorial n == n + sumtorial (n-1)
> -- Two errors
> -- * if n < 0 sumtorial doesn't stop (Hspec + QuickCheck)
> -- * Exection 
> --
> --     @
> --       Variable not in scope:
> --         wkName
> --           :: [Char]
> --              -> template-haskell-2.11.1.0:Language.Haskell.TH.Syntax.Name
> --     @
> -- To solve: read https://github.com/phadej/cabal-doctest
~~~~



CHANGELOG.md
# Revision history for yorgey-cis194

## 0.4.1.0 -- 2021-02-26 Homework 3, Wholemeal programming

### Modified

* src/HigerOrder.hs     -- Exercise 1
* CHANGELOG.md          -- Fix an error

## 0.4.1.0 -- 2021-02-25 Homework 3, Wholemeal programming

### Modified

* src/HigerOrder.hs      -- Exercise 1, statement

## 0.4.0.1 -- 2021-02-25 Homework 3

### Added

* src/HigerOrder.hs      -- Assignment 1, haddock structure

## 0.4.0.0 -- 2021-02-25 Homework 3

### Added

* doc/04-higher-order.lhs -- CIS 194 Week 4
* doc/04-higher-order.pdf -- CIS 194 Homework 4

## 0.3.0.3 -- 2020-12-21 Homework 3

### Added 

* doc/03-rec-poly.pdf   -- assignment
* src/Golf.hs           -- assignment source

### Changed

* CHANGELOG.md
* package.yaml          -- new version
* yorgey-cis194.cabal

## 0.3.0.2 -- 2020-11-23 RecPoly module

### Added

* src/RecPoly.lhs       -- module from lesson

### Changed

* CHANGELOG.md          -- fix last records
* package.yaml          -- new version
* yorgey-cis194.cabal
* doc/03-rec-poly.lhs   -- Changed and then restore from d78bee4 commit

## 0.3.0.1 -- 2020-11-03 03-rec-poly

### Added

* doc/03-rec-poly.lhs   -- downloaded lesson lhs

### Changed

New version

* CHANGELOG.md
* package.yaml
* yorgey-cis194.cabal

## 0.2.0.2 -- 2020-11-02 02-ADTs

### Added

Downloaded files from assignment

* data/error.log
* data/sample.log
* doc/02-ADTs.pdf
* src/Logs.hs           -- module downloaded for assignment
* src/ADTs.lhs    

Assignment

* src/LogAnalysis.hs    -- Assignment file

### Changed

New version

* CHANGELOG.m           -- new version
* package.yaml          -- added 'data-files'd
* yorgey-cis194.cabal
* TODO                  -- format of checklist

## 0.2.0.1 -- 2020-10-28 02-ADTs

Week 2: Algebraic Data Types

* doc/02-ADTs.lhs               -- downloaded lesson lhs file
* src/ADTs.lhs                  -- module version

### Changed 

Fix doctest

* src/Hanoi.hs                  -- Fix doctest and comment about pending implement `reeve`
* src/Cards.hs                  -- Fix doctest

Fix Hspec test

* test/cis194/HanoiSpec.hs      -- comment `reeve` test and use pendingWith instead

New version 

* CHANGELOG.md
* package.yaml
* yorgey-cis194.cabal

## 0.1.0.1 -- 2020-10-26 01-Intro

### Added
* doc/style.pdf
* doc/01-intro.lhs
* doc/images/pure.jpg
* doc/images/relax.jpg
* doc/images/static.jpg
* doc/images/function-machine.png
* doc/images/haskell-logo-small.png
* src/Intro.lhs
* test/cis194/Spec.hs
* test/cis194/IntroSpec.hs
* src/Cards.hs
* test/cis194/CardsSpec.hs
* app/cccn.hs

### Changed
* package.yaml
* yorgey-cis194.cabal
* CHANGELOG.md

## 0.1.0.0 -- 2020-10-07

First version. Skeleton of a project:

* .gitignore
* CHANGELOG.md
* LICENSE
* README.md
* Setup.hs
* hie.yaml
* package.yaml
* yorgey-cis194.cabal
* app/Main.hs
* src/MyLib.hs
* test/MyLibTest.hs


test/cis194/Spec.hs
-- file Spec.hs
{-# OPTIONS_GHC -F -pgmF hspec-discover #-}


test/doctest/Main.hs
-- Main for doctest
module Main (main) where

import Test.DocTest

main :: IO()
main = doctest ["-XOverloadedStrings", "src"]


test/cis194/CardsSpec.hs
-- |
-- Module      :  CardsSpec
-- Copyright   :  erlnow 2020 - 2030
-- License     :  BSD3
--
-- Maintainer  :  erlestau@gmail.com
-- Stability   :  experimental
-- Portability :  unknown
--
-- Tests and use case for functions in 'Cards' module.
-- 
-- Note to run tests:
--
-- @
--      $ cabal new-run spec
-- @
--
-- or in @ghci@:
--
-- @
--      $ cabal new-repl spec
--      ghci> hspec CardsSpec.spec
-- @

module CardsSpec (spec) where

import Test.Hspec
-- import Test.QuickCheck
import Cards

spec :: Spec
spec = do
  describe "Exercise 1" $ do
    it "toDigits 1234" $ do
      toDigits 1234 `shouldBe` [1,2,3,4]
    it "toDigitsRev 1234" $ do
      toDigitsRev 1234 `shouldBe` [4,3,2,1]
    it "toDigits 0" $ do
      toDigits 0 `shouldBe` ([] :: [Integer])
    it "toDigits (-17)" $ do
      toDigits (-17) `shouldBe` ([] :: [Integer])

  describe "Exercise 2" $ do
    it "doubleEveryOther [8,7,6,5]" $ do
      doubleEveryOther [8,7,6,5] `shouldBe` [16,7,12,5]
    it "doubleEveryOther [1,2,3]" $ do
      doubleEveryOther [1,2,3] `shouldBe` [1,4,3]

  describe "Exercise 3" $ do
    it "sumDigits [16,7,12,5]" $ do
      sumDigits [16,7,12,5] `shouldBe` 1 + 6 + 7 + 1 + 2 + 5

  describe "Exercise 4" $ do
    it "validate 4012888888881881" $ do
      validate 4012888888881881 `shouldBe` True
    it "validate 4012888888881882" $ do
      validate 4012888888881882 `shouldBe` False


src/Golf.hs
-- |
-- Module      :  Golf
-- Description :  Homework assignment 3
-- Copyright   :  erlnow 2020 - 2030
-- License     :  BSD3
--
-- Maintainer  :  erlestau@gmail.com
-- Stability   :  experimental
-- Portability :  unknown
--
-- Homework assignment 3 from Yorgey's CIS 194

module Golf where

import Data.List (transpose)

-- * Code Golf!
--
-- $hw
--
-- This assignment is simple: there are three tasks listed below. For each
-- task, you should submit a Haskell function with the required name and type
-- signature which accomplishes the given task and is /as short as possible/.
--
-- original pdf is placed in @doc/03-rec-poly.lhs@.

-- ** Exercise 1 Hopscotch
--
-- $ex1
--
-- Your first task is to write a function
--
-- @
--   skips :: [a] -> [[a]]
-- @
--
-- The output of 'skips' is a list of list. The first list in the output should
-- be the same as the input list. The second list in the output should contain
-- every second from the input list...and the /n/th list in the output should
-- contain every /n/th element from the input list.
--
-- For example:
--
-- >>> skips "ABCD"   == ["ABCD", "BD", "C", "D" ]
-- True
-- >>> skips "hello!" == ["hello!", "el!", "l!", "l", "o", "!"]
-- True
-- >>> skips [1]      == [[1]]
-- True
-- >>> skips []       == []
-- True
--
-- Solution:
--
-- If we have a function that returns every /n/th element of a list, given @n@
-- between @1@ and @length l@ and being @l@ the input list the solution is
-- straightforward. My solution is a list comprehension where from @1@ to the
-- length of @l@ I generate the element using my function to skip every @i@th
-- element of the input list.
--
-- 'every' drops @n-1@ elements and call itself recursively to choose next
-- /n/th element, building the list using the __cons__ operator @:@.

-- |function that expects a list and return a list of lists. The first element
-- is always the input list, the second element contains every second from the
-- input list, until the last element that contains the /n/th element.
skips :: [a] -> [[a]]
skips []  = []
skips xs  = [ every xs i | i <- [1..length xs]]

-- |returns a list formed by elements to the given list every @n@ positions.
every :: [a] -> Int -> [a]
every l n = case drop (n-1) l of
                (y:ys) -> y : every ys n
                []     -> []

-- ** Exercise 2 Local maxima
--
-- $ex2
--
-- /local maximum/ of a list is an element of the list which is strictly
-- greater than both the elements immediately before and after it. For
-- example, in the list @[2,3,4,1,5]@, the only local maximum is 4, since
-- it is greater than the elements immediately before and after it (3 and
-- 1). 5 is not a local maximum since there is no element that comes after
-- it.
--
-- Write a function
--
-- @
-- localMaxima :: [Integer] -> [Integer]
-- @
--
-- which finds all the local maxima in the input list and returns them in order.
-- For example:
--
-- >>> localMaxima [2,9,5,6,1]
-- [9,6]
-- >>> localMaxima [2,3,4,1,5]
-- [4]
-- >>> localMaxima [1,2,3,4,5]
-- []
--
-- Solution:
--
-- 'zip3' is a function that forms a list of 3-tuples given three lists. 'localMaxima'
-- constructs a list of three consecutive elements given a list of 'Integer's, using
-- this function. 
-- @
-- zip3 xs (drop 1 xs) (drop 2 xs)
-- @
--
-- This list is filtered using 'filter' function that return a list of element that
-- satisfies and argument.
--
-- @
-- filter (\(x,y,z) -> x < y && y > z)
-- @
--
-- At last, using map to return the second element of the 3-tuple that is the local
-- maximum.

-- |Returns a list of local maximum in a given list of 'Integer's.
localMaxima :: [Integer] -> [Integer]
localMaxima xs = map (\(_,y,_) -> y)
               . filter (\(x,y,z) -> x < y && y > z)
               $ zip3 xs (drop 1 xs) (drop 2 xs)

-- ** Exercise 3: Histogram
--
-- $ex3
--
-- For this task, write a function
--
-- @
-- histogram :: [Integer] -> String
-- @
--
-- which takes as input a list of 'Integer's between 0 and 9 (inclusive),
-- and outputs a vertical histogram showing how many of each number
-- where in the input list. You may assume that the input list does not
-- contain any numbers less than zero or greater than 9 (that is, it does
-- not matter what your function does if the input does contains such
-- numbers). Your output must exactly match the output shown in the examples
-- bellow.
--
-- >>> histogram [1,1,1,5]
--  *
--  *
--  *   *
-- ==========
-- 0123456789
-- 
-- >>> histogram [1,4,5,4,6,6,3,4,2,4,9]
--     *
--     *
--     * *
--  ******  *
-- ==========
-- 0123456789

-- |It counts the number of a given element in a list
count :: Eq a => a -> [a] -> Int
count x = length . filter (==x)

-- |Search elements of a given list in another list and returns how much
-- given elements appear in the second list as a list.
counts :: Eq a => [a] -> [a] -> [Int]
counts es xs = [count x xs | x <- es]

-- |Create a String with a number of "*"
asterisks :: Int -> [Char]
asterisks x = replicate x '*' ++ repeat ' '

-- til this point I can create a vertical histogram. How can I translate
-- to horizontal?
--
-- >>> putStrLn $ unlines $ map asteriks $ counts [0..9] [1,4,5,6,6,3,4,2,4,9] 
--
-- adding spaces to the final to every string

-- | print a horizontal histogram
-- histogram :: [Int] -> String
histogram xs = putStrLn $ bars ++ "----------\n0123456789\n"
  where bars = unlines
             . reverse
             . takeWhile (any (/= ' '))
             . transpose
             . map asterisks $ counts [0..9] xs


src/Hanoi.hs
-- |
-- Module      :  Hanoi
-- Description :  Home work assigment for Week 1: Towers of Hanoi
-- Copyright   :  erlnow 2020 - 2030
-- License     :  BSD3
--
-- Maintainer  :  erlestau@gmail.com
-- Stability   :  experimental
-- Portability :  unknown
--
-- Homework 1, exercise 5 Towers of Hanoi
-- 
-- Note to run tests:
--
-- @
--      $ cabal new-run spec
-- @
--
-- or in @ghci@:
--
-- @
--      $ cabal new-repl spec
--      ghci> hspec HanoiSpec.spec
-- @

module Hanoi (
             -- **Exercise 5
             -- $ex5
               Peg, Move, hanoi
             -- **Exercise 6
             -- $ex6
             , reeve
) where

-- $ex5
-- TODO: how to embed images in Haddock using local files? ![Towers of
-- Hanoi](hanoi.jpg)
--
-- The /Towers of Hanoi/ is a classic puzzle with a solution that can be
-- described recursively. Disk of different sizes are stacked on three pegs;
-- the goal is to get from a starting configuration with all disks stacked on
-- the last peg, as shown in Figure 1(TODO: add image)
--
-- The only rules are
--
-- * you may only move one disk at a time, and
-- * a larger disk may never be stacked on top of a smaller one.
--
-- For example, as the first move all you can do is move the topmost, smallest
-- disk onto a different peg, since only one disk may be moved at a time.
--
-- From this point, is /illegal/ to move to the configuration shown in Figure
-- 3(TODO: add image), because you are not allowed to put the green disk on top
-- of the smaller blue one.
--
-- To move \(n\) discs (stacked in increasing size) from peg \(a\) to peg \(b\)
-- using peg \(c\) as temporary storage,
--
-- 1.   move \(n - 1\) discs from \(a\) to \(c\) using \(b\) as temporary storage
-- 2.   move the top disc from \(a\) to \(b\)
-- 3.   move \(n - 1\) discs for \(c\) to \(b\) using \(a\) temporary storage.
--
-- For this exercise, define a function 'hanoi' with the following type:
--
-- @
-- type Peg = String
-- type Move = (Peg, Peg)
-- hanoi :: Integer -> Peg -> Peg -> Peg -> [Move]
-- @
--
-- Given the number of discs and the names for the three pegs, 'hanoi' should
-- return a list of moves to be perfomed to move the stack of disc form the
-- firs peg to the second.
--
-- Note that a type declaration, like @type Peg = String@ above, makes a /type
-- synonym/. In this case 'Peg' is declared as a synonym for 'String', and the
-- two names 'Peg' and 'String' can now be used interchangeably. Giving more
-- descriptive names to types in this way can be used to give shorted names to
-- complicated types, or (as here) simply to help with documentation.
--
-- Example:
--
-- >>> hanoi 2 "a" "b" "c" == [("a","c"),("a","b"),("c","b")]
-- True

-- |Name of peg
type Peg = String

-- |a move from one peg to other
type Move = (Peg, Peg)

-- |move @n@ discs from first peg to second using the last as temporary storage
hanoi :: Integer -> Peg -> Peg -> Peg -> [Move]
hanoi n a b c
  | n <= 0    = []
  | otherwise = (hanoi (n-1) a c b) ++ [(a,b)] ++ (hanoi (n-1) c b a)

-- $ex6
-- __(Optional)__ What if there are four pegs instead of three?  That is, the
-- goal is still move a stack of discs from th e first peg to the last peg,
-- whithout ever placing a larger disc on top of a smaller one, but now there
-- are two extra pegs that can be used as "temporary" storage instead of only
-- one. Write a function similar to 'hanoi' which solves this problem in as few
-- moves as posible.  It should be possible to do it in far fewer moves than
-- with three pegs. For example, with three pegs it takes \(2^15 - 1 = 32767)
-- moves to transfer 15 discs. With four pegs it can be done in 129 moves.
-- (See Exercise 1.17 in Graham, Knuth, and Patashnik, /Concrete Mathematics/,
-- second ed., Addison-Wesley, 1994)

-- |reeve's puzzle, like hanoi but with four pegs.
--
-- __TODO__: __undefined__. Pending implement.
reeve :: Integer -> Peg -> Peg -> Peg -> Peg -> [Move]
reeve = undefined


src/HigerOrder.hs
-- |
-- Module      :  HigerOrder
-- Description :  Home work assigment for Week 4: Higher Order Functions 
-- Copyright   :  erlnow 2020 - 2030
-- License     :  BSD3
--
-- Maintainer  :  erlestau@gmail.com
-- Stability   :  experimental
-- Portability :  unknown
--
-- CIS 194, Homework 4
-- 

-- Note to run tests:
-- @
--      $ cabal new-run spec
-- @
--
-- or in @ghci@:
--
-- @
--      $ cabal new-repl spec
--      ghci> hspec HanoiSpec.spec
-- @

module HigerOrder where

-- * Higer Order Functions
--
-- $text1
--
-- __What to turn in__: you should turn a single .hs (or .lhs), which
-- __must type check__.

-- ** __Exercise 1: Wholemeal programming__
--
-- $text2
--
-- Reimplement each of the following functions in a more idiomatic
-- Haskell style. Use /wholemeal programming/ practices, breaking
-- each function into a pipeline of incremental transformations to
-- an entire data structure. Name your functions @fun1'@ and @fun2'@
-- respectively.
--
-- @
--   fun1 :: [Integer] -> Integer
--   fun1 []       = 1
--   fun1 (x:xs)
--     | even x    = (x - 2) * fun1 xs
--     | otherwise = fun1 xs
--
--   fun2 :: Integer -> Integer
--   fun2 1             = 0
--   fun2 n | even n    = n + fun2 (n `div` 2)
--          | otherwise = fun2 (3*n + 1)
-- @
--
-- Hint: For this problem you may wish to use the functions
-- 'iterate' and 'takeWhile'. Look them up in the 'Prelude' documentation
-- to see what they do.

-- | fun1 function from exercise 1
fun1 :: [Integer] -> Integer
fun1 []       = 1
fun1 (x:xs)
 | even x    = (x - 2) * fun1 xs
 | otherwise = fun1 xs

-- | fun2 function from exercise 1
fun2 :: Integer -> Integer
fun2 1             = 0
fun2 n | even n    = n + fun2 (n `div` 2)
       | otherwise = fun2 (3*n + 1)

-- | /wholemeal programming/ version of 'fun1'
--
-- prop> fun1 xs == fun1' xs
fun1' :: [Integer] -> Integer
fun1' = product . map (subtract 2) . filter (even)

-- | /wholemeal programming/ version of 'fun2'
--
-- prop> fun2 n == fun2' n
fun2' :: Integer -> Integer
fun2' n = 
  let nextN m
       | even m    = m `div` 2
       | otherwise = 3*m + 1
  in sum . filter (even) . takeWhile (>1) $ iterate nextN n

-- ** __Exercise 2: Folding with trees__
-- 
-- $text3
--

-- ** __Exercise 3: More folds!__
--
-- $text4
--

-- ** __Exercise 4: Finding primes__
--
-- $text5
--


src/LogAnalysis.hs
{-# OPTIONS_GHC -Wall #-}
-- |
-- Module      :  LogAnalysis
-- Description :  CIS 194: Homework 2
-- Copyright   :  erlnow 2020 - 2030
-- License     :  BSD3
--
-- Maintainer  :  erlestau@gmail.com
-- Stability   :  experimental
-- Portability :  unknown
--
-- Yorgey's CIS 194: Introduction to Haskell (Spring 2013)

module LogAnalysis where

import Log

-- * Log file parsing
--
-- **Exercise 1
--
-- $ex1
--
-- The first step is figuring out how to parse an individual message. Define a
-- function:
-- 
-- @
--   parseMessage :: String -> LogMessage
-- @
--
-- which parse an individual line from the log file. For example:
--
-- @
--   parseMessage "I 29 la la la" == LogMessage Info 29 "la la la"
--
--   parseMessage "This is not in the right format" == Unknow "This is not in the right format"
-- @
--
-- Once we can parse one log message, we can parse a whole log file.
-- Define a function
--
-- @
--   parse :: String -> [LogMessage]
-- @
--
-- which parse an entire log file at once and returns its contents as a list of
-- 'LogMessage's.
--
-- To test your function, use the 'testParse' function provided in the "Log"
-- module, giving it as arguments your 'parse' function, the number of lines to
-- parse, and the log file to parse from (which should also be in the same
-- folder as your assignment). For example, after loading your assignment
-- into @GHCi@, type something like this at the prompt:
--
-- @
--   testParse parse 10 "error.log"
-- @
--
-- Don't reinvent the wheel! (That's so /last/ week.) Use "Prelude" functions
-- to make your solution as concise, high-level, and functional as possible.
-- For example, to convert a 'String' like @"562"@ into an 'Int', you can 
-- use the 'read' function. Other functions which may (or may not) be useful
-- to you include 'lines', 'words', 'unwords', 'take', 'drop' and '(.)'

-- |parse a 'String' and returns a value of type 'LogMessage'
--
-- First version: assume that all logs are well formed
parseMessage :: String -> LogMessage
parseMessage [] = Unknown ""
parseMessage xs = case hl of
                    "I" -> if getTime == Nothing
                              then Unknown xs
                              else LogMessage Info (j getTime) getMsg 
                    "W" -> if getTime == Nothing
                              then Unknown xs
                              else LogMessage Warning (j getTime) getMsg
                    "E" -> if getError == Nothing || getTime == Nothing
                              then Unknown xs
                              else LogMessage (Error (j getError)) (j getTime) getMsg
                    _   -> Unknown (xs)
                where
                  j (Just a) = a
                  line = words xs       -- has at last one word
                  hl = head line
                  msgError = head. tail $ line
                  msgTime  = if hl == "E"
                                then head . tail . tail $ line
                                else head . tail $ line 
                  getError = if null msgError
                                then Nothing
                                else Just (read msgError :: Int)
                  getTime  = if null msgTime
                                then Nothing
                                else Just (read msgTime :: Int)
                  getMsg   = if hl == "E"       -- can be "" 
                                then unwords . tail . tail . tail $ line
                                else unwords . tail . tail $ line 

-- |parse a full log file and returns its contents as a list of 'LogMessage's.
parse :: String -> [LogMessage]
parse [] = []
parse xs = map parseMessage (lines xs)

-- *Putting the logs in order
--
-- **Exercise 2
--
-- $ex2
--
-- Define a function
--
-- @
--   insert :: LogMessage -> MessageTree -> MessageTree
-- @
--
-- which insert a new 'LogMessage' into an existing 'MessageTree', producing a
-- new 'MessageTree'. 'insert' may assume that it is given a sorted
-- 'MessageTree', and must produce a new sorted 'MessageTree' containing the
-- new 'LogMessage' in addition to the contents of the original 'MessageTree'.
--
-- However, note that if 'insert' is given a 'LogMessage' which is 'Unknown',
-- it should return the 'MessageTree' unchanged.

-- |insert a 'LogMessage' in a sorted 'MessageTree' and return a new sorted
-- 'MessageTree'. If 'LogMessage' is 'Unknown' return the 'MessageTree'
-- unchanged.
insert :: LogMessage -> MessageTree -> MessageTree
insert (Unknown _) mt = mt
insert lm (Leaf)   = Node Leaf lm Leaf
insert l@(LogMessage _ t _) (Node i l'@(LogMessage _ t' _) d)
  = if t < t' then Node (insert l i) l' d
              else Node i l' (insert l d)

-- **Exercise 3
--
-- $ex3
--
-- Once we can insert a single 'LogMessage' into a 'MessageTree', we can build
-- a complete 'MessageTree' from a list of messages. Specifically, define a
-- function
--
-- @
--   build :: [LogMessages] -> MessageTree
-- @
--
-- which builds up a MessageTree containing the messages in the list, be
-- successively inserting the messages into a 'MessageTree' (beginning with a
-- 'Leaf')

-- |from a given list of 'LogMessage's builds up a sorted tree of 'LogMessages'
build :: [LogMessage] -> MessageTree
build = foldr insert Leaf 

-- **Exercise 4
--
-- $ex4
--
-- Finally, define the function
--
-- @
--   inOrder :: MessageTree -> [LogMessage]
-- @
--
-- which takes a sorted 'MessageTree' and produces a list of all the
-- 'LogMessages' it contains, sorted by time stamp from smallest to biggest.
-- (This is known as /in-order traversal/ of the MessageTree.)
--
-- With these functions, we can now remove 'Unknown' messages and sort the
-- well-formed messages using an expression such as:
--
-- @
--   inOrder (build tree)
-- @
--
-- [Note: there are much better ways to sort a list; this is just an exercise
-- to get you working with recursive data structures!]

inOrder :: MessageTree -> [LogMessage]
inOrder Leaf = []
inOrder (Node i l d) = inOrder i ++ [l] ++ inOrder d

-- *Log file postmortem
--
-- ** Exercise 5
--
-- $ex5
--
-- Now that we can sort the log messages, the only thing left to do is extract
-- the relevant information. We have decided that \"relevant\" means \"errors
-- with a severity of at least 50\".
--
-- Write a function
--
-- @
--   whatWentWrong :: [LogMessage] -> [String]
-- @
--
-- which take an /unsorted/ list of LogMessages, and return a list of the
-- messages corresponding to any errors with a severity of 50 or greater,
-- sorted by timestamp. (Of course, you can use your functions from the
-- previous exercises to do the sorting.)
--
-- For example, suppose our log file looked like this:
--
-- @
--   I 6 Completed armadillo processing
--   I 1 Nothing to report
--   E 99 10 Flange failed!
--   I 4 Everything normal
--   I 11 Initiating self-destruct sequence
--   E 70 3 Way too many pickles
--   E 65 8 Bad pickle-flange interaction detected
--   W 5 Flange is due for a check-up
--   I 7 Out for lunch, back in two time steps
--   E 20 2 Too many pickles
--   I 9 Back from lunch
-- @
--
-- This file is provided as @sample.log@. There are four errors, three of which
-- have a severity of greater than 50. The output of 'whatWentWrong' on
-- @sample.log@ ought to be
--
-- @
--   [ "Way too many pickles"
--   , "Bad pickle-flange interaction detected"
--   , "Flange failed!"
--   ]
-- @
--
-- You can test your 'whatWentWrong' function with 'testWhatWentWrong', which
-- is also provided by the "Log" module. You should provide 'testWhatWentWrong'
-- with your parse function, your 'whatWentWrong' function, and the name of the
-- log file to parse.

-- |a function that takes an unsorted list of LogMessage items and return a
-- list of string from the LogMessages sorted by timestamp that are Errors with
-- a severity greater than 50.
whatWentWrong :: [LogMessage] -> [String]
whatWentWrong = map getMsg . filter meaningLog . inOrder . build

meaningLog :: LogMessage -> Bool 
meaningLog (LogMessage (Error s) _ _) = s > 50 
meaningLog _ = False

getMsg :: LogMessage -> String
getMsg (Unknown s)        = s
getMsg (LogMessage _ _ s) = s

-- *Miscellaneous
--
-- $misc
--
-- * We will test your solution on log files other than the ones we have given
-- you, so no hardcoding!
--
-- * You are free (in fact, encouraged) to discuss the assignment with any of
-- your classmates as long as you type up your own solution.

-- *Epilogue
--
-- **Exercise 6
--
-- $ex6
--
-- For various reasons we are beginning to suspect that the recent mess was
-- caused by a single, egoistical hacker. Can you figure out who did it?)
-- 
whatWentWrong' :: [LogMessage] -> [(TimeStamp,String)]
whatWentWrong' = map getMsg' . filter meaningLog . inOrder . build
  where getMsg' (LogMessage _ t s) = (t,s)

huntHacker :: [LogMessage] -> [String]
huntHacker = map getMsg . filter errors . inOrder . build
  where errors (LogMessage (Error _) t _) = t < 131 
        errors _ = False


src/Cards.hs
-- |
-- Module      :  Cards
-- Description :  Home work assigment for Week 1: check credit card numbers
-- Copyright   :  erlnow 2020 - 2030
-- License     :  BSD3
--
-- Maintainer  :  erlestau@gmail.com
-- Stability   :  experimental
-- Portability :  unknown
--
-- For validate a Credit Card most providers rely on a checksum formula
-- for distinguishing valid numbers from random collections.
-- In this library 'validate' check if a number is a valid credit card
-- number.
--
-- This is the first assignment for week 1 of /Yorgey's cis194 course/.
--
-- The assignment can be read on
-- <https://www.seas.upenn.edu/~cis194/spring13/hw/01-intro.pdf>
-- 
-- Note to run tests:
--
-- @
--      $ cabal new-run spec
-- @
--
-- or in @ghci@:
--
-- @
--      $ cabal new-repl spec
--      ghci> hspec CardsSpec.spec
-- @

module Cards (
             -- **Exercise 1
             -- $ex1
               toDigits, toDigitsRev
             -- **Exercise 2
             -- $ex2
             , doubleEveryOther
             -- **Exercise 3
             -- $ex3
             , sumDigits
             -- **Exercise 4
             -- $ex4
             , validate
  ) where

-- $intro
-- For validate a /credit card number/:
--
-- * Double the value of every second digit beginning from the right.
--   That is, the last digit is unchanged; the second-to-last digit
--   is doubled; the third-to-last is unchanged; and so on. For example
--   [1,3,8,6] becomes [2,3,16,6]
-- * Add the digits of the double values and the undoubled digits from
--   the original number. For example, @[2,3,16,6]@ becomes
--   @2+3+1+6+6=18@.
-- * Calculate the remainder when the sum is divided by 10. For the
--   above example, the remainder would be 8.
--
-- If the result is 0, then the number is valid.

-- $ex1
-- We need to first find the digits of a number. Define the functions
--
-- @
-- toDigits    :: Integer -> [Integer]
-- toDigitsRev :: Integer -> [Integer]
-- @

-- |'toDigits' converts positive Integers to a list of digits. 'toDigitsRev'
-- should do the same, but with digit reversed.
--
-- >>> toDigits 1234 == [1,2,3,4]
-- True
-- >>> toDigitsRev 1234 == [4,3,2,1]
-- True
-- >>> toDigits 0 == []
-- True
-- >>> toDigits (-17) == []
-- True
toDigits :: Integer -> [Integer]
toDigits  = reverse . toDigitsRev

-- |Converts a reversed positive Integer to digits
toDigitsRev :: Integer -> [Integer]
toDigitsRev n | n <= 0 = []
              | otherwise = n `mod` 10 : toDigitsRev (n `div` 10)

-- $ex2
-- Once we have the digits in the proper order, we need to
-- double every other one. Define a function
--
-- @
-- doubleEveryOther :: [Integer] -> [Integer]
-- @

-- |Doubles every other number /beginning from the right/, that is, 
-- second-to-last, fourth-to-last, ... numbers are doubled.
--
-- >>> doubleEveryOther [8,7,6,5]
-- [16,7,12,5]
-- >>> doubleEveryOther [1,2,3]
-- [1,4,3]
doubleEveryOther :: [Integer] -> [Integer]
doubleEveryOther xs = (reverse . doubleEveryOther' . reverse) xs 
  where doubleEveryOther' []       = []
        doubleEveryOther' [x]      = [x]
        doubleEveryOther' (x:y:ns) = x : 2*y : doubleEveryOther' ns

-- $ex3
-- The output of doubleEveryOther has a mix of one-digit
-- and two-digit numbers. Define the function
--
-- @
-- sumDigits :: [Integer] -> Integer
-- @
--
-- to calculate the sum of all digits.
-- 
-- Example:
--
-- >>> sumDigits [16,7,12,5] == 1 + 6 + 7 + 1 + 2 + 5
-- True

-- |Calculates the sum of all digits of a list of Integers.
--
-- Note: The list has numbers with one digit and two digits.
-- Numbers with two-digits are the doubles of one digit number
sumDigits :: [Integer] -> Integer
sumDigits ns = sum [if x > 9 then sum2dig x else x | x <- ns]
                      where sum2dig n = n `div` 10 + n `mod` 10
          -- previously I have write:
          -- @sum [if x > 9 then x `mod` 9 else x | x <- ns]@
          -- but failed when @x == 9@ or @x == 18@, in this case
          -- @9 `mod` 9 == 0@ or
          -- @18 `mod` 9 == 0@
          -- the first case i don't know why, @9 > 9 == False@
          -- the second case comes from doubleEveryOther and
          -- i had not think this case

-- $ex4
-- Define the function
--
-- @
-- validate :: Integer -> Bool
-- @
--
-- that indicates whether an 'Integer' could be a valid credit
-- card number. This will use all functions defined in the previous
-- exercises.
-- Examples
--
-- >>> validate 4012888888881881
-- True
--
-- >>> validate 4012888888881882
-- False

-- |Indicates whether an 'Integer' could be a valid credit
-- card number.
validate :: Integer -> Bool
validate n = (sumDigits . doubleEveryOther . toDigits) n `mod` 10 == 0
